libafl/src/stages/calibrate.rs

```diff
@@ -0,0 +1,213 @@
+use crate::{
+    bolts::current_time,
+    corpus::{Corpus, PowerScheduleTestData},
+    executors::{Executor, HasObservers},
+    fuzzer::Evaluator,
+    inputs::Input,
+    observers::{MapObserver, ObserversTuple},
+    stages::Stage,
+    state::{HasCorpus, HasMetadata},
+    Error,
+};
+use alloc::string::{String, ToString};
+use core::{marker::PhantomData, time::Duration};
+use num::Integer;
+use serde::{Deserialize, Serialize};
+
+/// The default mutational stage
+#[derive(Clone, Debug)]
+pub struct CalibrationStage<C, E, EM, I, O, OT, S, T, Z>
+where
+    T: Integer + Default + Copy + 'static + serde::Serialize + serde::de::DeserializeOwned,
+    C: Corpus<I>,
+    E: Executor<EM, I, S, Z> + HasObservers<I, OT, S>,
+    I: Input,
+    O: MapObserver<T>,
+    OT: ObserversTuple<I, S>,
+    S: HasCorpus<C, I> + HasMetadata,
+    Z: Evaluator<E, EM, I, S>,
+{
+    map_observer_name: String,
+    stage_max: usize,
+    #[allow(clippy::type_complexity)]
+    phantom: PhantomData<(C, E, EM, I, O, OT, S, T, Z)>,
+}
+
+// The number of times we run the program in the calibration stage
+const CAL_STAGE_MAX: usize = 8;
+
+impl<C, E, EM, I, O, OT, S, T, Z> Stage<E, EM, S, Z>
+    for CalibrationStage<C, E, EM, I, O, OT, S, T, Z>
+where
+    T: Integer + Default + Copy + 'static + serde::Serialize + serde::de::DeserializeOwned,
+    C: Corpus<I>,
+    E: Executor<EM, I, S, Z> + HasObservers<I, OT, S>,
+    I: Input,
+    O: MapObserver<T>,
+    OT: ObserversTuple<I, S>,
+    S: HasCorpus<C, I> + HasMetadata,
+    Z: Evaluator<E, EM, I, S>,
+{
+    #[inline]
+    #[allow(clippy::let_and_return)]
+    fn perform(
+        &mut self,
+        fuzzer: &mut Z,
+        executor: &mut E,
+        state: &mut S,
+        manager: &mut EM,
+        corpus_idx: usize,
+    ) -> Result<(), Error> {
+        let iter = self.stage_max;
+        let handicap = state
+            .metadata()
+            .get::<PowerScheduleStats>()
+            .ok_or_else(|| Error::KeyNotFound("PowerScheduleStats not found".to_string()))?
+            .queue_cycles;
+
+        // Timer start
+        let start = current_time();
+
+        for _i in 0..iter {
+            let input = state
+                .corpus()
+                .get(corpus_idx)?
+                .borrow_mut()
+                .load_input()?
+                .clone();
+            let (_, _) = fuzzer.evaluate_input(state, executor, manager, input)?;
```

Comment:
here in AFL we use run_target that is just a plain run without looking at the bitmap. Dunno if we should follow the same design here, maybe the user will assume that observers pre and post exec hooks will be always executed and so in that case it is good to use evaluate_input. On the other hand, it is slower. Put a comment so we can remember that we differ from AFL.

Reply:
ah, yes I can use run_target, it's faster :>

Reply:
At first I wanted to use evaluate_input, because aflpp checks the checksum of the bitmap, and if the fuzzer has found anything interesting, the calibration would run last longer.
https://github.com/AFLplusplus/AFLplusplus/blob/32a0d6ac31554a47dca591f8978982758fb87677/src/afl-fuzz-run.c#L427
but I just felt it would be slower.